### PR TITLE
Machine Gun Bug Fix

### LINKED ISSUE #556: M56D Cade Exploit
## Description
You can jump over cades without opening them while using the M56D.

## Difference between expected and actual behavior
It should stop you from jumping the cades.

## Steps to reproduce
1. Deploy a M56D right to a cade (doesn't matter if it has barbed wire or not)
2. Walk in the direction of the casde that the M56D is attached too.
3. Exploit achieved.

## Issue bingo (not required to be all checked)

- [X] Issue could be reproduced at least once
- [X] Issue happened in a recent (less than 7 days ago) round
- [X] Couldn't find an existing issue about this (https://github.com/cmss13-devs/cmss13/issues)

### PR DESCRIPTION
**Pull request by lajolico**
## About The Pull Request

Removed the ability to vault over barricades (tables, cades, anything) or access the m56d MG from the outside of cades, both are considered game exploits. Players can no longer rotate their character while firing the machine gun, as this is not consistent with how you fire an actual MG. Players can enter a Machine gun(m56d/m2c) from behind (where the handles are) or to the left or right of that tile, refer to the image below: 
image

When the player exits a machine gun, they animate out, instead of just "walking" out either to the left, back, or right of the machine gun.

When a player "enters" a machine gun, they can either resist() out of the machine gun or press the UI button (Exit MG) located at the top left of their screen. 
image

Fixes #556 

## Why It's Good For The Game

The current state of the machine guns is that you can vault over cades inside or outside if you are adjacent to an m56d, with the exception of the m2c as it has anti-cade code.   You shouldn't be able to walk out from either direction while leaving a machine gun. This PR fixes that. Further, when "exiting" out of the machine gun, you simply walk out or under the MG. With this PR, you will animate out cleanly, adding a nice aesthetic. Also, when you're manning an MG, you're manning it, you're not "dancing" or shooting it backward, which doesn't make sense. Case in point:
image

## Changelog

:cl: John_Rico
add: Added a UI button to exit a MG or you can resist 
add: Added an animation "exiting" an M2c or m56d w/ variant
fix: Can no long vault over cades or man a MG on the other side of a barricade
/:cl:

### CODE REVIEW

File: code/modules/cm_marines/smartgun_mount.dm [hunk 1]
```diff
@@ -1468,30 +1487,32 @@
...
+			else
+				if(user.interactee) //Make sure we're not manning two guns at once, tentacle arms.
+					to_chat(user, "You're already manning something!")
+					return
 
-		if(user.get_active_hand() == null && user.get_inactive_hand() == null)
-			user.set_interaction(src)
+			if(user.get_active_hand() == null && user.get_inactive_hand() == null)
+				user.freeze()
+				user.set_interaction(src)
```
[the remark this answers is not in the document]
Comment (lajolico) [reply code/modules/cm_marines/smartgun_mount.dm:1510]: You are able to move still and jump over cades during the animation, so I kept user.freeze() in there.

File: code/modules/mob/living/carbon/human/human_abilities.dm [hunk 2]
```diff
@@ -596,3 +596,23 @@ CULT
...
+	. = ..()
+	UnregisterSignal(M, COMSIG_MOB_MG_ENTERED)
+
+/datum/action/human_action/mg_exit/action_activate()
+	if(!can_use_action())
+		return
+
+	var/mob/living/carbon/human/H = owner
+	SEND_SIGNAL(H, COMSIG_MOB_MG_EXIT)
+	remove_from(H)
```
[the remark this answers is not in the document]
Comment (lajolico) [reply code/modules/mob/living/carbon/human/human_abilities.dm:618]: Fixed

File: code/modules/cm_marines/smartgun_mount.dm [hunk 3]
```diff
@@ -827,64 +832,77 @@
...
 	user.visible_message(SPAN_NOTICE("[user] mans \the [src]."),SPAN_NOTICE("You man \the [src], locked and loaded!"))
-	user.update_canmove()
 	user.forceMove(src.loc)
 	user.setDir(dir)
 	user_old_x = user.pixel_x
 	user_old_y = user.pixel_y
 	user.reset_view(src)
 	update_pixels(user)
 	operator = user
-	user.unfreeze()
```
[the remark this answers is not in the document]
Comment (lajolico) [reply code/modules/cm_marines/smartgun_mount.dm:871]: Fixed

File: code/modules/cm_marines/smartgun_mount.dm [hunk 4]
```diff
@@ -827,64 +832,77 @@
...
 		operator = null
+	remove_action(user, /datum/action/human_action/mg_exit)
+
+	UnregisterSignal(user, list(
+		COMSIG_MOB_MG_EXIT,
+		COMSIG_MOB_RESISTED
+	))
+
+	user.update_canmove()
+	user.unfreeze()
```
[the remark this answers is not in the document]
Comment (lajolico) [reply code/modules/cm_marines/smartgun_mount.dm:905]: Did not know that. Fixed!

**Comment by fira** [comment]
Makes sense to me, would like to see it testmerged for a few rounds if that's still applicable

**Comment by fira** [comment]
You can consider these "add" in changelog code/fix/qol probably, add is for features, don't waste your GBPs

**Comment by lajolico** [comment]
@fira Okay, thank you! Sounds good, would love to see this in action.

**Comment by fira** [comment]
Testmerged this for ~20 hours, seems to be going well. Let's go.

**Pull request merged**